### ISSUE TITLE
Tag search results on threads v messages

### ISSUE THREAD

**Issue by hugoroy** [astroidmail/astroid#376]
Let's say I receive and email and manually add tag: topic/x

The thread following that email grows, but new messages do not automatically get the tag topic/x 

If I search for tag:topic/x the thread won't show up in results where I'd expect, i.e. taking into account recent messages in the thread.

This seems like unexpected to me. Isn't it?

On my phone but will try to be more descriptive if that's not enough.

**Comment by yeled** [comment]
@hugoroy I know what you're talking about, don't worry. I've tried to make a small test suite for this a long time ago but got bored :)

**Comment by mxmehl** [comment]
I know that problem, too. I have a more folder-based approach, so I wrote a script which:

- uses a predefined list of used tags (I have ~20-30 which I often use)
- loops through this list and tags all threads with this tag if one of its messaged have this tag
  - given that these messages are located in INBOX
- moves these newly tagged messages to its predefined folder

Quite complicated but works well for my setup. If you're interested, [have a look at it](https://src.mehl.mx/mxmehl/mail-config/src/master/astroid/scripts/move-tag.sh) and the corresponding files.

**Comment by mrvdb** [comment]
If I recall correctly, the `killed-threads` filter in `afew` implements this exact idea, albeit for the purpose of ignoring the threads, might be worth looking at for inspiration.

**Comment by gauteh** [comment]
There has been some discussions on the notmuch list on how to search 
for threads (rather than messages). Essentially the search in notmuch 
works on messages, not threads, and what you want is for the search 
terms to be matched on the whole thread - not on a single message.

There is some talk on implementing searching 'thread-documents' or just 
making all the search terms into 'OR's which should put the best match 
at the top.

e.g.: id:1497594312.ahjx44w937.astroid@strange.none
      id:1471858269.x2m28lgosh.astroid@strange

Propagating tags works for some things (like muting, killed), but not so 
good for other stuff like 'flagged' or 'unread', etc..

**Comment by hugoroy** [comment]
Hi Gaute, thanks everyone for your answers so far.

Since search and tags are core features, I think it's important to get
that right : -)

↪ Gaute Hope / août 3, 2017 11:19:
> There has been some discussions on the notmuch list on how to search 
> for threads (rather than messages). Essentially the search in notmuch 
> works on messages, not threads, and what you want is for the search 
> terms to be matched on the whole thread - not on a single message.

As Astroid's current view is thread-based, maybe it needs some work
so that the results feel right.

> There is some talk on implementing searching 'thread-documents' or just 
> making all the search terms into 'OR's which should put the best match 
> at the top.
> 
> e.g.: id:1497594312.ahjx44w937.astroid@strange.none
>       id:1471858269.x2m28lgosh.astroid@strange

I don't understand.

> Propagating tags works for some things (like muting, killed), but not so 
> good for other stuff like 'flagged' or 'unread', etc..

How does propagating tags work?

If I tag a thread with "mute" all incoming messages in the thread will
also get the "mute" tag, or?



* * * * *


Here's the use case that prompted me to start this discussion.


Alice works on Project X. Alice gets an email (E1) from somebody that
relates to X, so in the Astroid thread view, Alice adds the tag
`project/x`.

The thread goes on, Alice gets replies (E2, E3), etc. In the Astroid
thread view, the thread is correctly identified with the tag `project/x`
but incoming messages (E2, E3) added to the thread don't get the tag
`project/x` automatically.

Now, some time passes. Alice needs to get to step 2 of Project X and
wants to search for past emails on Project X. So Alice naturally
launches a search in Astroid for `tag:project/x`.

It seems to me that the search results will not take into account E2 and
E3, but only E1.

So in the user's perspective, here's what's right: the thread does show
up in the results, so Alice will find E1, E2, and E3.

And here what does not feel right: E1's date will be taken into account,
instead of the last message (E3) in the thread, so the thread shows up
below and is harder to find.

When a project goes on, this starts to be really difficult.



For search results in a thread view mode, I would expect that Astroid
considers ”threads” as the result for any message in the thread matching
the search query.


Does that make sense? I've tried to make this clear but I'm not sure I
managed : -)

**Comment by hugoroy** [comment]
Sorry, I meant thread index in the previous message, not thread view. :-)

**Comment by mxmehl** [comment]
I have brought up this issue before because I wasn't aware that notmuch searches message-based. You explained it much better, so I just reference #346

**Comment by hugoroy** [comment]
I see your point there, is what's relevant I think:

> I learnt that if I search for tag:XX, notmuch sorts the threads for the newest message with this tag, and not the newest message of the whole thread. Because the sent mails weren't tagged with XX, only the received mails were tagged and considered for sorting.

**Comment by hugoroy** [comment]
So it looks like we need to implement something at notmuch's level?

**Comment by gauteh** [comment]
Yes, thread-search must be implemented at notmuch level (ref. notmuch mailing list message-ids I referred to)

**Comment by hugoroy** [comment]
Related, there's already interesting things about "muted" and thread as results for search here:
https://notmuchmail.org/faq/

> Can I tag threads?
>
> No. Tagging is message based.

> 
> It is possible, however, to make tags propagate to all messages in a thread using a little bit of scripting in the post-new hook. For example, to add the muted tag to all messages in threads that have at least one message with the muted tag:
> 
>     THREAD_TAGS="muted"
>     for tag in "$THREAD_TAGS"; do
>             notmuch tag +$tag $(notmuch search --output=threads tag:$tag)
>     done
> 
> You can add other tags to THREAD_TAGS as needed. Note that this is one way only; you need to explicitly remove the tag from all the messages in a thread to stop it from propagating again.
>
